tests/chainer_tests/test_optimizer.py

```diff
@@ -11,6 +11,167 @@
+        self.assertEqual(hook.call_count, 1)
+
+        args = hook.call_args_list[0][0]
+        self.assertEqual(len(args), 2)
+        self.assertIs(args[0], self.update_rule)
+        self.assertIs(args[1], self.var)
+
+    def test_remove_hook(self):
+        hook = mock.MagicMock()
+        self.update_rule.add_hook(hook, name='hook')
+        self.update_rule.remove_hook('hook')
+        self.update_rule.update(self.var)
+        self.assertEqual(hook.call_count, 0)
+
+    def test_add_hook_with_function_name(self):
+        hook_body = mock.MagicMock()
+
+        def foo(update_rule, data, grad):
+            hook_body(update_rule, data, grad)
+
+        self.update_rule.add_hook(foo)
+        self.update_rule.remove_hook('foo')
+        self.update_rule.update(self.var)
+        self.assertEqual(hook_body.call_count, 0)
+
+    def test_add_hook_no_name(self):
+        class CallableWithoutName(object):
+            def __call__(self, update_rule, param):
+                pass
+
+        with self.assertRaises(ValueError):
+            self.update_rule.add_hook(CallableWithoutName())
+
+    def test_add_hook_duplicated_name(self):
+        self.update_rule.add_hook(mock.MagicMock(), name='foo')
+        with self.assertRaises(ValueError):
+            self.update_rule.add_hook(mock.MagicMock(), name='foo')
+
+    def test_remove_hook_not_exist(self):
+        with self.assertRaises(KeyError):
+            self.update_rule.remove_hook('foo')
+
+    def test_disabled_update_rule(self):
+        self.update_rule.update_core = mock.MagicMock()
+        self.update_rule.enabled = False
+        self.update_rule.update(self.var)
+        self.assertEqual(self.update_rule.update_core.call_count, 0)
+
+        self.update_rule.enabled = True
+        self.update_rule.update(self.var)
+        self.assertEqual(self.update_rule.update_core.call_count, 1)
+
+    def setup_state(self):
+        def init_state(data):
+            state = self.update_rule.state
+            state['a'] = 0
+            state['b'] = np.array([1, 2, 3], dtype=np.float32)
+        self.update_rule.init_state = init_state
+
+    @attr.gpu
+    def test_state_copy_to_gpu(self):
+        self.setup_state()
+
+        def update_core(param):
+            self.assertIsInstance(self.update_rule.state['a'], int)
+            self.assertIsInstance(self.update_rule.state['b'], cuda.ndarray)
+
+        self.update_rule.update_core = update_core
+        self.var.to_gpu()
+        self.update_rule.update(self.var)
+
+    @attr.multi_gpu(2)
+    def test_state_copy_to_another_gpu(self):
+        self.setup_state()
+
+        def update_core(param):
+            self.assertIsInstance(self.update_rule.state['b'], cuda.ndarray)
+            self.assertEqual(self.update_rule.state['b'].device.id, 1)
+
+        self.update_rule.update_core = lambda param: None
+        self.update_rule.update(chainer.Variable(
+            cuda.to_gpu(self.data), grad=cuda.to_gpu(self.grad)))
```

Comment:
This `update` lets the update rule to copy the state to GPU 0. This method actually tests if the second `update` correctly copies the state from GPU 0 to GPU 1.